File: 2411-spiral-matrix-iv/spiral-matrix-iv.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        vector<vector<int>> ans(m, vector<int>(n, -1));
        int i=0, j=0, curr_direction = 0, movement[4][2] = {{0,1}, {1,0}, {0,-1}, {-1, 0}};
        while (head != nullptr) {
            ans[i][j] = head->val;

            int newi = i + movement[curr_direction][0];
            int newj = j + movement[curr_direction][1];

            if (min(newi, newj) < 0 || newi >= m || newj >= n || ans[newi][newj] != -1) {
                curr_direction = (curr_direction + 1) % 4;
            }
            i += movement[curr_direction][0];
            j += movement[curr_direction][1];

            head = head->next;
        }
        return ans;
    }
};

File: 2411-spiral-matrix-iv/spiral-matrix-iv.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        i, j = 0, 0
        movement = [[0,1],[1,0],[0,-1],[-1,0]]
        curr_dir = 0
        ans = [[-1 for _ in range(n)] for _ in range(m)]

        while head is not None:
            print(str(i) + " " + str(j) + '\n')
            ans[i][j] = head.val

            new_i = i + movement[curr_dir][0]
            new_j = j + movement[curr_dir][1]

            if (min(new_i, new_j) < 0 or new_i >= m or new_j >= n or ans[new_i][new_j] != -1):
                curr_dir = (curr_dir + 1) % 4

            i += movement[curr_dir][0]
            j += movement[curr_dir][1]

            head = head.next
        
        return ans